### ISSUE TITLE
feat: Keep track of `state_dict` corresponding to best model according to validation accuracy

### ISSUE THREAD

**Issue by PaulScemama** [awslabs/fortuna#93]
This may be already possible but I can't seem to figure out how to do it. Is there a way to use the `FitConfig` components to keep track of the best `state_dict` of a model according to validation accuracy and then saving that state dict? 

Thanks so much!

**Comment by gianlucadetommaso** [comment]
Hi Paul,
in general, you can use `FitMonitor` to keep track of metrics during the training, and enable early stopping. Furthermore, you can use `FitCheckpointer` to define how many checkpoints you want to store in your directory, and how often do you want to save.

If you enable early stopping, e.g. you set `early_stopping_patience=2` in the `FitMonitor`, and you pass a `save_checkpoint_dir` to the `FitCheckpointer`, then you are going to save a checkpoint by default every time the metric that you are tracking is improved. By default, early stopping tracks the validation loss, but you can configure `early_stopping_monitor` and set it to what you need.

Does this answer your question?

**Comment by PaulScemama** [comment]
I think so!

So say I don't want early stopping...:

```python
num_epochs = 100

optimizer=FitOptimizer()

monitor=FitMonitor(
    metrics=(accuracy,),
    early_stopping_patience=num_epochs,
    early_stopping_monitor="val_loss",
    eval_every_n_epochs=1,)


checkpointer=FitCheckpointer(
    save_checkpoint_dir="/checkpoints/"
    keep_top_n_checkpoints=1)

```

Will this code only save 1 checkpoint, and will this 1 checkpoint be the checkpoint where the model had the best validation loss? 

Pretty much I always want the checkpoint where the model did the best according to some validation metric. However I don't want to have to do early stopping if I don't want to. 
```

**Comment by gianlucadetommaso** [comment]
Like this it should be the last, since the final state is always saved if you pass a `save_checkpoint_dir`. But you should be able to do what you want if you set `keep_top_n_checkpoint=2`. Then the best checkpoint should be the second last, unless the last checkpoint is not only the last but also the best. In the latter case, there is gonna be only one checkpoint in the directory, corresponding to both the last and the best.

**Comment by PaulScemama** [comment]
Okay thank you! 

In the documentation it says `keep_top_n_checkpoints` is the "Number of past checkpoint files to keep." Maybe I'm missing something but according to that documentation it is not entirely obvious that setting `keep_top_n_checkpoints=2` would give you the *best* checkpoint as the second last (or the last if that is also the best). 

Maybe have something like: `keep_best_n_checkpoints` and have a mechanism that contains the best `n` checkpoints seen thus far and then writes those to file. What do you think?

**Comment by gianlucadetommaso** [comment]
Yes good points, I'll definitely try to improve this part, thanks. I'm working on a refactoring that also involves checkpointing. I'll take this into account! 👍🏼
